Src/Support/Google.Apis/Services/BaseClientService.cs

```diff
@@ -236,6 +261,39 @@ protected virtual BackOffHandler CreateBackOffHandler()
             return new BackOffHandler(new ExponentialBackOff());
         }
 
+        /// <summary>
+        /// Gets the effective URI taking into account the <see cref="UniverseDomain"/>.
+        /// </summary>
+        /// <param name="explicitUri">An explicit URI. May be null.</param>
+        /// <param name="defaultUri">A default URI. May be null.</param>
+        /// <returns>
+        /// <list type="bullet">
+        /// <item><paramref name="explicitUri"/> if not null.</item>
+        /// <item>
+        /// Otherwise, if <see cref="UniverseDomain"/> is not null, the result of replacing
+        /// <see cref="DefaultUniverseDomain"/> with <see cref="UniverseDomain"/>
+        /// in <paramref name="defaultUri"/>.
+        /// </item>
+        /// Otherwise <paramref name="defaultUri"/>.
+        /// </list>
+        /// </returns>
+        protected internal string GetEffectiveUri(string explicitUri, string defaultUri)
```

Comment:
@jskeet this is the method that we'll be using from generated code to calculate the base URI and the batch URI. I'll send that PR out tomorrow.